Add QueryInterceptor to presto-jdbc

robbypete opened the pull request:
Introduce chainable [query interceptors](https://dev.mysql.com/doc/connector-j/8.0/en/connector-j-interceptors.html) to Presto JDBC. QueryInterceptors can modify the behavior of a query or perform side effects. QueryInterceptors' `preProcess()` and `postProcess()` methods are run before and after the query is executed in the server, respectively. If more than one QueryInterceptor classname is provided separated by a semicolon, each will be called one-by-one from left to right in the classnames provided in the connection string. Both methods can optionally return a PrestoResultSet to override the ResultSet returned by the query.

Addresses issue #15142 

Test plan - Run new tests in TestQueryInterceptor, TestPrestoDriverUri, and TestJdbcConnection

```
== RELEASE NOTES ==

JDBC changes
* Add the interface `QueryInterceptor` to allow for custom logic to be executed before or after query execution. (:pr:`15565`)

tdcmeehan commented on presto-jdbc/src/main/java/com/facebook/presto/jdbc/AbstractConnectionProperty.java:
```diff
@@ -190,6 +193,34 @@ private StringMapConverter() {}
+            try {
+                return (QueryInterceptor) Class.forName(interceptor).getDeclaredConstructor().newInstance();
+            }
+            catch (Throwable e) {
+                System.err.print("Could not load QueryInterceptor class");
+                e.printStackTrace();
+                return null;
+            }
```
```suggestion
            catch (Throwable e) {
                throw new IllegalArgumentException(format("Could not load QueryInterceptor classes from %s", value), e);
            }
```

tdcmeehan commented on presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoConnection.java:
```diff
@@ -201,6 +206,9 @@ public void close()
                 try (PrestoStatement statement = new PrestoStatement(this)) {
                     statement.internalExecute("ROLLBACK");
                 }
+                for (QueryInterceptor queryInterceptor : this.queryInterceptorInstances) {
```
I think for autocommit queries, this won't have a transaction ID.  So it could be done outside of the if check above.  Also, we should take care to handle exceptions.  Something like this:

https://github.com/prestodb/presto/blob/04138bc26a4fc801630fe936463d89d071d00c0a/presto-common/src/main/java/com/facebook/presto/common/block/ClosingBlockLease.java#L66-L84

tdcmeehan commented on presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoConnection.java:
```diff
@@ -113,6 +115,9 @@
         timeZoneId.set(TimeZone.getDefault().getID());
         locale.set(Locale.getDefault());
+
+        this.queryInterceptorInstances = uri.getQueryInterceptors();
```
Let's do a defensive copy just to be safe.
```suggestion
        this.queryInterceptorInstances = ImmutableList.of(uri.getQueryInterceptors());
```

robbypete replied:
Unless I'm missing something, it looks like we can't easily cast a LinkedList to an ImmutableList. Would it make more sense to make the return type of ClassListConverter in AbstractConnectionProperty an ImmutableList instead of a LinkedList? We don't expect to modify it after it's built and I don't think we're seeing many benefits to using a LinkedList for our usage.

tdcmeehan replied:
My suggestion had a typo, it should be:

```java
this.queryInterceptorInstances = ImmutableList.copyOf(uri.getQueryInterceptors());
```

tdcmeehan commented on presto-jdbc/src/main/java/com/facebook/presto/jdbc/QueryInterceptor.java:
```diff
@@ -0,0 +1,25 @@
+
+public interface QueryInterceptor
+{
+    QueryInterceptor init(PrestoConnection conn, Map<String, String> props);
```
```suggestion
    QueryInterceptor init(PrestoConnection connection, Map<String, String> properties);
```

tdcmeehan commented on presto-jdbc/src/test/java/com/facebook/presto/jdbc/BaseQueryInterceptor.java:
```diff
@@ -0,0 +1,40 @@
+public class BaseQueryInterceptor
+        implements QueryInterceptor
+{
+    public QueryInterceptor init(PrestoConnection conn, Map<String, String> props)
```
```suggestion
    public QueryInterceptor init(PrestoConnection connection, Map<String, String> properties)
```

tdcmeehan reviewed (commented):
Some nits

mayankgarg1990 commented on presto-jdbc/src/main/java/com/facebook/presto/jdbc/QueryInterceptor.java:
```diff
@@ -0,0 +1,25 @@
+
+public interface QueryInterceptor
+{
+    QueryInterceptor init(PrestoConnection connection, Map<String, String> properties);
+    void destroy();
+    PrestoResultSet preProcess(String sql, Statement interceptedStatement);
+    PrestoResultSet postProcess(String sql, Statement interceptedStatement, PrestoResultSet originalResultSet);
```
We should add some documentation to these functions about the arguments and expected behavior since people in open source might want to implement these and not having to read the actual implementation will be helpful to quickly onboard people and move ahead with their business logic.

robbypete replied:
I added some documentation to these methods

mayankgarg1990 reviewed (commented):
I spent a bit more time thinking about the basic design of this PR and I have some concerns over how the implementation is done. 

First of all - I am guessing that you have a basic motive in mind that you want to solve - can you share that so that I can make sure that my further comments considers that.

In the existing design

1) In the pre-step we are delegating the execution of the query to the interceptor. This generally involves the interceptor calling `connection.getStatement().execute()` and this in turn invokes the whole chain of interceptors again internally and then you again call them externally. In certain situations, you can get into an infinite loop as a result of this - for example - in your `testMultiPreProcess` if we change the second processor to change the execution to convert `456` to `123` which is the opposite of the first one - I would generally just expect this to nullify the effect of first one, rather than getting into an infinite loop.

2) Something similar in the post step - someone can potentially just discard the results and completely re-run the query.

Depending on the use case you are trying to solve, can we start simple - maybe just modify the `sql` or some of the parameters of the query like `SESSION_PROPERTIES` ?

robbypete commented:
The use case that I specifically have for a QueryInterceptor is pretty narrow (which I think you picked up on): I want to be able to modify the sql in the client before it's sent the server. In #15142 Tim suggested to look into MySQL's implementation of the StatementInterceptor as inspiration and I thought their implementation of an interceptor is a good way for me to provide a more generic solution rather than just a sql filter for my own use case. 

> Depending on the use case you are trying to solve, can we start simple

There are other use cases which others are hoping to use a QueryInterceptor for which involve being able to perform side effects before/after query execution related to privacy and security for Tableau (and other corp apps using Presto). Starting simple might get complicated if we have to modify an interface down the road to expand on the interceptor concept as more use cases are developed. 

Admittedly, these use other use cases I'm trying to cover don't involve reusing the connection like I am in my tests so maybe a good path forward is to rethink how `preProcess` should act, like maybe it could act as a filter on the statement instead of allowing a ResultSet to be returned.

> this in turn invokes the whole chain of interceptors again internally and then you again call them externally. In certain situations, you can get into an infinite loop as a result of this

My thinking behind this design is that this is intended - I am reusing the same connection in the QueryInterceptor and all queries executed using that connection are treated the same. MySQL deals with this by using a `executeTopLevelOnly` flag in the QueryInteceptor interface where a QueryInterceptor can be marked to only run in the top-level query and not in queries issued by QueryInterceptors. We could either introduce that flag or, maybe better, make all QueryInterceptors top-level by default and instead make the flag signal whether a QueryInterceptor should run on all queries.

> Something similar in the post step - someone can potentially just discard the results and completely re-run the query.

This is intended. I think allowing for the modification of the response is in-line with what someone would expect from an Interceptor/Filter.

All that being said I'm open to ideas to making this fit in better and not introduce anything unexpected.

robbypete commented:
Added flag to QueryInterceptor signaling if the QueryInterceptor should intercept nested queries i.e. queries executed by a QueryInterceptor.

robbypete commented:
I did some refactoring and wrote more tests. 

I uncovered an issue after writing the test `testForceNestedIntercept()`. If an interceptor has `forceNestedIntercept()` return true and the interceptor returns a PrestoResultSet in each postProcess call (edge case, for sure) then the `checkArgument` in `addInnerResultSet()` throws (in PrestoResultSet).

A few ideas I have in mind are to either restrict queries when forceNestedIntercept() is true, restrict queries when the statementDepth counter is above 2, or to change how we handle inner ResultSets.

tdcmeehan commented:
My preference would be to remove `forceNestedIntercept` and have this always be disabled.  If we need to introduce it later, it's a backwards compatible change, and given that it's hard to think of non-esoteric use cases for it, it seems like skirting the issue for now is the quickest and most straightforward path.

robbypete commented:
Removed the forceNestedIntercept flag and changed logic so that only top-level queries are intercepted.

mayankgarg1990 commented:
My apologies for the delayed response - due to internal beginning of year processes I am a bit swamped - I will get to this PR early next week

tdcmeehan commented on presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoConnection.java:
```diff
@@ -602,6 +627,24 @@ public int getNetworkTimeout()
 
+    public void incrementStatementDepth()
+    {
+        int currentDepth = this.statementDepth.get();
+        this.statementDepth.set(currentDepth + 1);
```
```suggestion
        statementDepth.getAndIncrement();
```

tdcmeehan commented on presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoConnection.java:
```diff
@@ -602,6 +627,24 @@ public int getNetworkTimeout()
+
+    public void decrementStatementDepth()
+    {
+        int currentDepth = this.statementDepth.get();
+        checkArgument(currentDepth > 0, "Statement depth cannot go below 0");
+        this.statementDepth.set(currentDepth - 1);
```
```suggestion
        int currentDepth = statementDepth.get();
        checkArgument(currentDepth > 0, "Statement depth cannot go below 0");
        statementDepth.decrementAndGet();
```

tdcmeehan commented on presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoConnection.java:
```diff
@@ -80,6 +83,7 @@
     private final AtomicReference<Integer> networkTimeoutMillis = new AtomicReference<>(Ints.saturatedCast(MINUTES.toMillis(2)));
     private final AtomicReference<ServerInfo> serverInfo = new AtomicReference<>();
     private final AtomicLong nextStatementId = new AtomicLong(1);
+    private final AtomicInteger statementDepth = new AtomicInteger(0);
```
I think we should move this state into `PrestoStatement`.  A connection can create multiple statements, all of which can be executed independently and in different threads.  On the other hand, there is always just one statement per result set, so the state really should apply to just the statement.

robbypete replied:
Yeah that's what I was afraid of. If we put the depth inside the statement, wouldn't we lose that depth if the QueryInterceptor creates another Statement? I can test it to make sure.

tdcmeehan reviewed (changes requested):
Mostly LGTM.  Mainly I think the statement depth state should go to the statement (see below).

robbypete commented:
Moved the statementDepth into the Statement and removed the Connection parameter from the init.

tdcmeehan reviewed (commented):
So it looks like we now don't take in a `Connection` in the `init` method, this trivially avoids the problem of overlapping interceptors.  This forces one to use the `Statement` to execute a new query.

I think this is fine, however in order to better future proof this feature, I think we need to plug a hole in our current `PrestoResultSet`.  According to the JavaDoc for `Statement`:

> By default, only one ResultSet object per Statement object can be open at the same time. Therefore, if the reading of one ResultSet object is interleaved with the reading of another, each must have been generated by different Statement objects. All execution methods in the Statement interface implicitly close a statement's current ResultSet object if an open one exists.

There's similar language in the `ResultSet` JavaDoc: https://docs.oracle.com/javase/6/docs/api/java/sql/ResultSet.html#close()

Right now, an execution method will override the `currentResultSet` in the `Statement`, but not close the old one.  I think we need to also make sure the old result set is closed first.  This will alter the behavior for `postProcess`.  For example, one may no longer be able to "zip" two ResultSets (the first ResultSet will automatically be closed once we obtain the second ResultSet), but it's better to enforce this now before backwards incompatible behavior is added in the client.

```java
    private void clearCurrentResults()
            throws SQLException
    {
        PrestoResultSet rs = currentResult.getAndSet(null);
        if (rs != null) {
            rs.close();
        }
        currentUpdateCount.set(-1);
        currentUpdateType.set(null);
        currentWarningsManager.set(Optional.empty());
    }
```

robbypete commented:
Applied suggested changes to BaseQueryInterceptor.java and clearCurrentResults()

mayankgarg1990 commented on presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoConnection.java:
```diff
@@ -741,6 +770,31 @@ WarningsManager getWarningsManager()
+        return processInterceptors(interceptor -> interceptor.preProcess(sql, interceptedStatement), null);
+    }
+
+    PrestoResultSet invokeQueryInterceptorsPost(String sql, Statement interceptedStatement, PrestoResultSet originalResultSet)
+    {
+        return processInterceptors(interceptor -> interceptor.postProcess(sql, interceptedStatement, originalResultSet), originalResultSet);
+    }
```
in this function - we are passing a fixed reference to the originalResultSet rather than the one returned by the previous post processor - in my opinion we should ensure that the correct and latest result set is being passed

mayankgarg1990 reviewed (changes requested):
Overall the PR looks reasonable to me know. Thanks for addressing my concerns :) I do have a couple of comments around the code styling.

I have one thought that I want to discuss which is when is `close()` called on `PrestoResultSet`. The reason this `close()` is important is that it is linked to the execution of the underlying query and it stops the query if close is invoked. As of now the close is invoked in 2 locations - 

1) There is a chaining of result sets happening via the `innerResultSet` parameter in `PrestoResultSet` and the close is invoked in a chaining fashion when it is invoked on the last result set.
2) Close is called in `clearCurrentResults` which is called every time `executeInternal` is invoked. This should happen whenever one of the interceptor decides to invoke `execute` on the `PrestoStatement` object

Given that we have (2), do you feel that we need (1). One situation that we need (1) for sure is when we don't reuse `Statement` that is passed in rather just create a new JDBC connection and return a `ResultSet` created from that statement - but not sure if we are trying to optimize for that situation.

robbypete commented on presto-jdbc/src/main/java/com/facebook/presto/jdbc/QueryInterceptor.java:
```diff
@@ -0,0 +1,63 @@
+     * @param sql the SQL string of the query.
+     * @param interceptedStatement the Statement being executed.
+     * @param originalResultSet the intercepted ResultSet
+     * @return optional PrestoResultSet to be returned instead of the original PrestoResultSet
```
It's the same approach. I think the implementation in the BaseQueryInterceptor threw you off. In the default implementation in the interface I changed it to just return null.

mayankgarg1990 replied:
@robbypete - This code was the part that made me write this comment - 

```
if (shouldIntercept) {
                    resultSet = connection().invokeQueryInterceptorsPost(sql, this, resultSet);
                    verifyNotNull(resultSet, "invokeQueryInterceptorsPost should never return a null ResultSet");
                }
```

As I read it again carefully, it should be fine since the `processInterceptors` will just return the original ResultSet here.

robbypete commented:
I think you're right about `innerResultSet` and it can be removed since we're constraining the interceptors to the same Statement.

mayankgarg1990 commented:
Also - we should fix the release note to be more descriptive

1) Remove the `General changes` section
2) Rewrite it maybe as 

```
JDBC changes
* Add the interface `QueryInterceptor` to allow for custom logic to be executed before or after query execution. (:pr:`15565`)
```

robbypete commented:
Applied suggested changes.

- Updated `internalExecute()` so that ResultSets from top level queries are set in currentResult before calling `invokeQueryInterceptorsPost` which allowed me to remove innerResultSet logic
- Removed the BaseQueryInterceptor and put default methods in the interface
- Moved the destroy() calls so that it would still be called after failed `ROLLBACK` 
- Updated java docs
- Changed the return type of `preProcess` and `postProcess` to be `Optional<PrestoResultSet>`
- and more smaller changes

mayankgarg1990 commented on presto-jdbc/src/main/java/com/facebook/presto/jdbc/QueryInterceptor.java:
```diff
@@ -0,0 +1,71 @@
+     *
+     * @param properties the session properties supplied in the URI
+     */
+    void init(Map<String, String> properties);
```
thinking a bit more - how do you feel about having a default empty implementation for this as well. The caller can choose to implement it or ignore the - the interface itself does not require an implementation to implement this function.

robbypete replied:
Good point, I like it

mayankgarg1990 reviewed (approved):
looks good to me - last few comments more around code suggestions - so good to go from my side

mayankgarg1990 commented on presto-jdbc/src/main/java/com/facebook/presto/jdbc/QueryInterceptor.java:
```diff
@@ -0,0 +1,71 @@
+public interface QueryInterceptor
+{
+    /**
+     * Called once per {@link <PrestoConnection>} at instantiation.
```
You can just say `{@link PrestoConnection}`

robbypete commented on presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java:
```diff
@@ -236,25 +238,48 @@ final boolean internalExecute(String sql)
             // check if this is a query
-            if (client.currentStatusInfo().getUpdateType() == null) {
+            if (intercepted || client.currentStatusInfo().getUpdateType() == null) {
+                currentResult.set(resultSet);
```
Previously (without innerResultSet logic), if an interceptor's postProcess() executes a query using the intercepted Statement and returns a ResultSet then we'll lose the reference to the top level query's ResultSet and it will stay opened past the lifetime of the Statement. Adding it here means it can be cleared in clearCurrentResults() when the interceptor's postProcess() executes another query.

mayankgarg1990 replied:
Makes sense ! Should we move the second `set()` on line 283 to be inside the `if` condition then ?

robbypete replied:
Ah of course

robbypete commented:
Applied changes. Noticed that, because we got rid of the BaseQueryInterceptor, I had to update `testUriWithQueryInterceptors` in `TestPrestoDriverUri` and add a class that it could use in the test.

tdcmeehan reviewed (approved):
LGTM
